src/source/CppGenerator.scala

```diff
@@ -74,8 +75,9 @@ class CppGenerator(spec: Spec) extends Generator(spec) {
           (w: IndentWriter) => {
             w.wl("template <>")
             w.w(s"struct hash<$fqSelf>").bracedSemi {
-              w.w(s"size_t operator()($fqSelf type) const").braced {
-                w.wl("return std::hash<int>()(static_cast<int>(type));")
+              w.wl(s"using underlying_type = typename std::underlying_type<$fqSelf>::type;")
+              w.w(s"std::size_t operator()(const $fqSelf &type) const").braced {
```

Comment:
well, it is not a problem to get them either by value or reference, for I understood that the type will be `int` at least for a while and copying `int`s around won't hurt anybody. I'm going to rollback it in a few minutes.